### PR TITLE
feat(gui-app): sweep a Task's landed worktrees from History and the epic sidebar

### PR DESCRIPTION
**Pull request by tanveergill**
## What

Adds **Sweep**: a per-Task cleanup that removes the worktrees a Task owns once its work has landed. Two surfaces:

- a hover action + context-menu item on **History** rows;
- a new always-visible **task-status strip** in the open epic's sidebar (which also becomes the home for the History pin toggle).

Entirely client-composed: **no new host RPC and no protocol change**. Rows come from `worktree.listAllForHost`; removal streams through the existing `worktree.deleteByPath` cleanup runner, so the host-side busy-check and teardown semantics are unchanged.

## Act-time proof, not the cached listing

The host serves resolved rows indefinitely with `forceRefresh: false` (steady-state liveness is manual-refresh-owned). So a worktree that was clean+landed when last probed, then edited from an external terminal, would still read "proven safe" from cache — and `deleteByPath` force-removes a dirty tree, since its busy-check only covers Traycer-registered owners.

Opening the dialog therefore re-proves: a cheap un-probed walk finds the Task's paths, then **one** selection-mode call with `forceRefresh: true` re-derives disk facts and merge proofs for exactly those rows — bounded by the Task's worktree count, never the fleet.

Known residual (under-claims, never false-greens): PR facts are read non-blocking on the host, so the first forced probe after an *external* merge can still serve the stale `open` fact while the background `gh` re-probe lands; the row starts unchecked until the dialog is reopened.

## The dialog

Lists **every** Task-owned worktree with its branch, tier pill, PR chips and path — the same detail Settings ▸ Worktrees shows.

- Rows the shared classifier proves safe (`merged` "Landed" / `at-base-commit`, exclusively owned, not busy) **start checked**.
- Everything else starts **unchecked but stays checkable**, with an amber hint naming the specific loss (uncommitted changes, unmerged commits, an unmerged submodule branch, or a binding another Task still holds).
- Busy and still-resolving rows are **disabled** — the host would refuse them anyway.

Sweeping runs in the **background**, like Settings worktree deletion: confirm hands the run off and closes the dialog, a kickoff toast acknowledges it, and a summary toast reports the removed/failed tally once every stream settles.

## Two staleness fixes ride along

Both are reachable from any delete flow, not just Sweep:

- **Composer intents are purged on completion.** Staged worktree picks are deliberately never re-validated by the seeding tiers, so without this a pick staged before a sweep kept offering the deleted worktree as an existing-worktree intent in new chats.
- **Browse/launch pickers hide host-proven-missing rows** (terminal creation, file tree, git diff), *except* each surface's own current selection. Binding rows outlive worktree deletion by design so an owning chat can recover, but on other surfaces a long-dead row is unusable noise.

> Note: the picker cleanup only fully takes effect alongside a companion host-side fix (deleted paths currently never resolve, so they linger as "checking" rather than "missing"). That fix ships separately in the internal repo.

## Testing

- New: `use-epic-sweep-worktree-candidates-query.test.tsx` (act-time probe wire shape, per-row classification, the stale-clean → dirty regression, failure ⇒ no candidates), `purge-removed-worktree-intents.test.ts` (intent purge across both stores + the resolved-missing row filter).
- Updated: `file-tree-workspace-picker.test.tsx` — one pre-existing case asserted the old "always show missing rows" contract; it now covers both hiding and the selection exemption.
- `bun run compile`, ESLint, and the affected Vitest suites pass; `react-doctor` reports nothing on the new files.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    // the kickoff and reports the outcome via toasts; re-opening while the
+    // run is live keeps Sweep disabled so the same paths can't double-stream.
+    sweepMutation.mutate({
+      worktrees: checkedRows.map((row) => ({
+        worktreePath: row.entry.worktreePath,
+        branch: row.entry.branch,
+      })),
+    });
+    onOpenChange(false);
+  };
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx:121]: Fixed in d4f0ea03. You're right that `sweepMutation.isPending` is per-mount and the dialog is rendered independently by the History row and the task-status strip, so a run started from one surface was invisible to the other.

`useSweepingWorktreePaths()` now reads the in-flight paths from the mutation cache filtered by `epicMutationKeys.sweepWorktrees()` — the same shared-key pattern `usePendingSetPinnedEpicIds` uses for exactly this reason. Those rows are unchecked and disabled in every mounted dialog.

One deliberate narrowing: the guard is per-path, not global. A sweep of a *different* Task's worktrees leaves this dialog fully usable; only the paths actually streaming are locked.

File: clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+          <span
+            className="mt-1 block max-w-full truncate font-mono text-ui-xs text-muted-foreground"
+            title={entry.worktreePath}
+          >
+            {entry.worktreePath}
+          </span>
+          <SweepRowHint row={row} />
+        </span>
+      </label>
+    </li>
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx:331]: Fixed in d4f0ea03 — confirmed real. `WorktreePrPills` renders `WorktreePrAnchor`, an `<a>` with an `onClick` that opens the PR externally, so nesting it inside the row `<label>` meant clicking a PR link also toggled the checkbox (and interactive content inside a label is invalid markup).

The row is restructured: the `<li>` is now the flex container, the `<label htmlFor>` wraps only the branch text, and the pills/tier pill sit outside it. The checkbox keeps its `aria-label`, so the accessible name is unchanged.

File: clients/gui-app/src/lib/worktree/removed-worktree-refs.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Whether a staged/remembered folder intent references a removed worktree and
+ * can therefore no longer materialize: an `import` of a removed directory, an
+ * existing-branch checkout of a deleted branch, or a new-branch fork whose
+ * SOURCE branch was deleted. Branch matching is by name only (no repo
+ * qualification): a cross-repo name collision at worst over-purges a
+ * remembered default, which just re-seeds to the standard new-worktree choice
+ * - while under-purging would keep offering a selection that can never be
+ * created.
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/lib/worktree/removed-worktree-refs.ts]: Fixed in d4f0ea03. I'd noted the name-only matching as an accepted trade-off in the original comment, but you're right that it's the wrong call: on a cross-repo epic, sweeping `feat/x` in one repo silently purged a remembered `feat/x` in another.

`RemovedWorktreeRefs.branches` is now `ReadonlyArray<{ repoIdentifier, branch }>` and matching requires both the branch name and the repo identity. The sweep targets carry `entry.repoIdentifier` through to the purge.

Edge case kept deliberately conservative: when `repoIdentifier` is `null` on either side (no parseable origin), identity can't *rule out* a match, so the name comparison stands alone. New test covers both directions.

File: clients/gui-app/src/stores/worktree/worktree-intent-staging-store.ts [hunk 4]
```diff
@@ -362,6 +375,36 @@ export const useWorktreeIntentStagingStore =
...
+              changed = true;
+              if (entries.length === 0) {
+                delete intentByKey[id];
+                delete suspendedWorkspacePathsByKey[id];
+              } else {
+                intentByKey[id] = { entries };
+              }
+              // Bumped like every other slot write so a rejected in-flight
+              // action can't restore the just-purged selection.
+              revisionByKey = incrementStagingRevision(revisionByKey, id);
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/stores/worktree/worktree-intent-staging-store.ts:420]: Fixed in d4f0ea03 — real inconsistency with the neighbouring code. `unstageEntry` already prunes `suspendedWorkspacePathsByKey` for the entry it removes; my partial-purge branch only cleared it when the slot emptied entirely, so a surviving slot kept fail-closed metadata for entries that were just purged, which would block a later restage of that workspace.

The partial branch now filters suspended paths down to the surviving entries' workspace paths, deleting the key when nothing remains.

File: clients/gui-app/src/components/epic-canvas/sidebar/task-status-strip.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        >
+          {/* Optimistic pin state - the icon flips at click time; the brief
+              disabled window only serializes rapid re-toggles. */}
+          <Pin className={cn("size-3.5", pinned === true && "fill-current")} />
+        </button>
+      </TooltipTrigger>
+      <TooltipContent>{label}</TooltipContent>
+    </Tooltip>
+  );
+}
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epic-canvas/sidebar/task-status-strip.tsx]: Not applying this one — the no-spinner pin is a deliberate, documented pattern rather than an oversight.

`HistoryPinControl` in `epics-list-panel.tsx` carries the rationale verbatim: "The pin state is optimistic - it flips at click time - so the icon always shows the row's current state; the brief disabled window only serializes rapid re-toggles, with no spinner." `useEpicSetPinned` is optimistic (`onMutate` patches the caches), so there is no pending *state* to represent — the icon is already showing the post-click truth. Adding a spinner here would make the strip's pin behave differently from the History row's for the same mutation.

The guideline's pending recipe applies to the actions that genuinely await a result; Sweep's confirm button follows it.

File: clients/gui-app/src/components/epics/epics-list-panel.tsx [hunk 6]
```diff
@@ -1391,6 +1427,84 @@ function historySelectedForDelete(args: {
...
+          onClick={props.sweep.requestSweep}
+        >
+          <Paintbrush />
+        </Button>
+      </TooltipTrigger>
+      <TooltipContent>Sweep this task's worktrees</TooltipContent>
+    </Tooltip>
+  );
+}
+
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epics/epics-list-panel.tsx:1483]: False positive — `group-hover:` is correct here, and it's what the sibling action already uses.

The row card carries **both** group names: `historyRowCardClassName` returns `"group relative min-w-0 flex-1 rounded-md ..."` (bare `group`), while the outer `<li>` carries `group/list-row`. The sweep button is rendered inside the row card, immediately before `{deleteControl}` — and `HistoryRowDeleteControl` uses `group-hover:opacity-100` for exactly the same reveal. So the unnamed `group-hover:` matches the card, and both hover actions appear together.

Switching to `group-hover/list-row:` would actually change behaviour: it would key the reveal to the outer `<li>` (which includes the selection-checkbox gutter) rather than the card, diverging from delete.

File: clients/gui-app/src/hooks/epic/use-epic-sweep-worktree-candidates-query.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        epicId ?? "",
+      ),
+      queryFn: fetchFreshTaskWorktreesNormalized,
+      enabled: epicId !== null && readiness.isReady,
+      // Always stale: every dialog open re-runs the forced probe rather than
+      // trusting a previous open's proof.
+      staleTime: 0,
+      retry: false,
+    }),
+  );
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktree-candidates-query.ts:135]: False positive for this app — focus/reconnect refetches are already off globally.

`createAppQueryClient()` in `src/lib/query-client.ts` sets `refetchOnWindowFocus: false` and `refetchOnReconnect: false` in `defaultOptions.queries`, and this query overrides neither. So `staleTime: 0` only means "re-run when a new observer mounts" — i.e. on dialog open, which is precisely the act-time proof this hook exists to provide. Refocusing the window while the dialog is open will not flip it back to "Checking worktrees…".

[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktree-candidates-query.ts:135]: Same as the sibling thread on this hook: `createAppQueryClient()` already pins `refetchOnWindowFocus: false` and `refetchOnReconnect: false` app-wide (`src/lib/query-client.ts`), and this query doesn't override them. With no focus/reconnect refetch, `staleTime: 0` costs nothing extra — it only guarantees a fresh forced probe when the dialog mounts, which is the intended behaviour.

File: clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        openStreamTransport,
+        hostId,
+        variables.worktrees.map((target) => target.worktreePath),
+      );
+      return { ...outcome, hostId };
+    },
+    onSuccess: (result, variables) => {
+      emitSweepSummaryToast(result.removed.length, result.failed.length);
+      purgeIntentsForRemovedWorktrees(variables.worktrees, result.removed);
+      invalidateWorktreeListingAndBindingCaches(queryClient, result.hostId);
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts]: Not changing this — the host is already captured once and used consistently, and the guideline's mechanism doesn't apply to a bare `useMutation`.

`mutationFn` reads `client.getActiveHostId()` **once** at the top, passes that value to `runWorktreeCleanup`, and returns it on the result; `onSuccess` then invalidates with `result.hostId`. So the sweep and its cache invalidation are guaranteed to target the same host — a swap mid-run cannot redirect either.

The `onMutate`-context pattern in the guideline is what `useHostMutation` provides for single-RPC mutations. This isn't one: the whole mutation is a streamed multi-path run, so it uses a bare `useMutation`, and TanStack does not pass `onMutate`'s context into `mutationFn`. Capturing at the top of `mutationFn` is the equivalent guarantee — if anything it reads a *fresher* host than `onMutate` would, since it runs later.

[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts:96]: Leaving this for a follow-up rather than this PR, deliberately.

`invalidateWorktreeListingAndBindingCaches` is a **shared** helper: the Task-delete cleanup (`useEpicBatchDelete`) and the Settings delete flow call the same function and have the identical exposure. Adding `cancelQueries` only on the Sweep path would leave the two flows behaving differently for the same hazard, and changing the shared helper alters the delete flows' behaviour, which wants its own validation and belongs in a change scoped to that helper.

Worth noting the window is small here: the invalidation runs in `onSuccess` after every delete stream has settled, and the listing scope is `refetchType: "active"`, so mounted observers refetch immediately against post-delete disk truth. Happy to open a separate issue for adding cancellation to the shared helper across all three call sites.

File: clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    },
+    onError: (error) => {
+      toast.error(error.message);
+    },
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts:100]: Not applicable here — `toastFromHostError` takes a `HostRpcError` (`src/lib/host-error-toast.ts`: `toastFromHostError(error: HostRpcError, fallback: string)`), and this mutation's error channel never carries one.

There's no host RPC in this path. `runWorktreeCleanup` drives `worktree.deleteByPath` **streams** and resolves every path into `removed`/`failed` rather than rejecting — per-path failures are reported by the summary toast (`reportableWarningToast`, with the reportable context), not by `onError`. The only value that reaches `onError` is the plain `Error` thrown by the `hostId === null` guard, whose message is already user-facing ("No host connection - can't sweep worktrees.").

Routing that through `toastFromHostError` would be a type error and would run `HostRpcError`-specific logic (unauthorized suppression, fatal-error notification) against an error that has none of that shape.

File: clients/gui-app/src/lib/worktree/removed-worktree-refs.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ * means the repo could not be identified (no parseable origin), so identity
+ * cannot be used to RULE OUT a match - the name comparison then stands alone,
+ * which is the conservative direction for a purge.
+ */
+function repoMatches(
+  left: RemovedBranchRepo | null,
+  right: RemovedBranchRepo | null,
+): boolean {
+  if (left === null || right === null) return true;
+  return left.owner === right.owner && left.repo === right.repo;
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/lib/worktree/removed-worktree-refs.ts:53]: Fixed in bf73cf81 (now 62904f4f after the rebase) — you're right that the blanket null-fallback was wrong, though I landed on a narrower rule than "both known or nothing".

The problem with requiring both identities known is that it silently reintroduces the bug this purge exists to fix. A local repo with no parseable origin has `repoIdentifier: null` on BOTH the removed worktree and its own intents, so "both must be known" would purge nothing there — the swept worktree would keep being offered as an "existing worktree" in new chats forever, which is exactly the report that prompted this code.

So the rule is now three-way:
- **both known** → must be equal (your scenario: `acme/app`'s `feat/x` can no longer touch `acme/other`'s)
- **both null** → branch name decides (the local-no-origin case, which must still purge)
- **one of each** → **no match** — an unidentifiable side cannot be shown to be the same repo, and guessing there costs a valid user selection

That closes the case you flagged (removed=null vs intent=known → no longer purges) without giving up the case the feature depends on.

On your "carry another stable repository key" suggestion — I looked for one and the wire doesn't have it: `worktreeHostEntrySchema` carries `worktreePath`, `repoLabel` and `repoIdentifier`, but no source workspace path, so the intent's `workspacePath` has nothing to join against. Adding one would be a protocol change; the three-way rule gets the same safety without it.

Two new tests cover both directions of the half-identified case.

File: clients/gui-app/src/hooks/epic/use-epic-sweep-worktree-candidates-query.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktree-candidates-query.ts:135]: Not applying this — `useTabHostId()` would throw at this call site, and the app-wide client is the correct scope here.

`<TabHostProvider>` wraps **tiles**, not Epic chrome: the only production mount is `epic-canvas/renderers/tile-render.tsx`, and `useTabHostId()` throws outside it ("must be called inside <TabHostProvider>… if you need the renderer-default host, use useReactiveActiveHostId()"). `EpicSweepAction` lives in the Epic status row, rendered by `EpicShell` straight off the route — above every tile — so a tab-scoped read has nothing to read.

It is also the established scope for Task-level worktree data. `useTaskWorktreeMetadata` (History rows), `useTaskDeleteWorktreeCandidates` (Task-delete dialog) and Settings ▸ Worktrees all resolve through `useHostClient()`; the sweep hooks match them, and the same dialog is rendered from History, which is unambiguously app-scoped.

The domain rule this cites is about chat/terminal **artifacts** bound to a host for life — the tiles. A Task's worktree bindings aren't tile state; they're per-host records read on the default host connection.

File: clients/gui-app/src/lib/worktree/worktree-row-resolved-missing.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * selection so a worktree deleted while selected stays visible (as its
+ * disabled badge) instead of silently vanishing out from under the user.
+ */
+export function withoutResolvedMissingRows(
+  rows: ReadonlyArray<WorktreeBindingSelectorRowV12>,
+  keepRunningDir: string | null,
+): ReadonlyArray<WorktreeBindingSelectorRowV12> {
+  return rows.filter(
+    (row) => !isResolvedMissingRow(row) || row.runningDir === keepRunningDir,
+  );
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/lib/worktree/worktree-row-resolved-missing.ts:48]: Fixed in 57191ac8 — good catch. Rows are keyed `(hostId, runningDir)` and my exemption compared the path alone, so with two hosts sharing a `runningDir` the non-selected host's dead row survived the filter.

`withoutResolvedMissingRows` now takes a `SelectedWorktreeRef { hostId, runningDir }` and matches on both. All three call sites pass their selection's host: git-diff from `selectedRepo.hostId`, file-tree from `props.hostId` (null-guarded with `selectedPath`), terminal from `explicitRow.hostId`.

File: clients/gui-app/src/stores/worktree/__tests__/purge-removed-worktree-intents.test.ts [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+  it("keeps a resolved-missing row that is the surface's current selection", () => {
+    const missing = row({
+      runningDir: "/wt/gone",
+      disabledReason: "missing_worktree_path",
+    });
+    expect(withoutResolvedMissingRows([missing], "/wt/gone")).toEqual([
+      missing,
+    ]);
+  });
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/stores/worktree/__tests__/purge-removed-worktree-intents.test.ts:313]: Added in 57191ac8: "exempts only the selected host's row when two hosts share a runningDir" — two resolved-missing rows on the same `runningDir` across `host-1`/`host-2`, asserting only the selected host's row survives. It fails against the old path-only identity.

File: clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+  /** A sweep of this exact path is already streaming (from any surface). */
+  readonly isSweeping: boolean;
+  readonly interactionDisabled: boolean;
+  readonly onToggle: (worktreePath: string, checked: boolean) => void;
+}) {
+  const { row, checked, isSweeping, interactionDisabled, onToggle } = props;
+  const entry = row.entry;
+  const branch = entry.branch ?? "detached HEAD";
+  const disabled = row.disabled || interactionDisabled || isSweeping;
+  const checkboxId = `sweep-worktree-${entry.worktreePath}`;
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epics/sweep-worktrees-dialog.tsx]: Fixed in 57191ac8 — real bug on Windows paths. The id is now `useId()`, so it can never contain a space and the `htmlFor` association holds regardless of the path.

File: clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+      const count = variables.worktrees.length;
+      toast.info(
+        `Sweeping ${count} worktree${count === 1 ? "" : "s"} in the background…`,
+      );
+    },
+    mutationFn: async (variables) => {
+      const hostId = client.getActiveHostId();
+      if (hostId === null) {
+        throw new Error("No host connection - can't sweep worktrees.");
+      }
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/hooks/epic/use-epic-sweep-worktrees-mutation.ts]: Fixed in 57191ac8. The kickoff toast moved out of `onMutate` and into `mutationFn`, after the `hostId === null` guard — a disconnected host now shows only the error toast instead of "Sweeping N worktrees…" immediately contradicted by "No host connection". Updated the hook's doc accordingly.

File: clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx [hunk 15]
```diff
@@ -36,6 +36,11 @@ function renderPill(state: EpicSyncPillState) {
   return render(pillTree());
 }
 
+function pillClaimsSynced(): boolean {
+  const pill = screen.queryByTestId("epic-connection-pill");
+  return pill?.getAttribute("aria-label") === "All changes synced";
+}
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx:50]: Fixed in 57191ac8 — you're right that it could pass vacuously. `pillClaimsSynced()` now reads through `screen.getByRole("button")`, so a missing pill throws instead of quietly satisfying the `false` assertions in the syncing/connected cases.

File: clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx [hunk 16]
```diff
@@ -48,17 +53,20 @@ describe("<EpicConnectionPill />", () => {
...
 
-    expect(screen.getByTestId("epic-connection-pill").textContent).toBe(
-      "All changes synced",
-    );
+    // Icon-only in the steady state: the claim moved off the visible copy and
+    // onto the accessible name + tooltip, so the status row stays compact.
+    expect(screen.getByTestId("epic-connection-pill").textContent).toBe("");
+    expect(
+      screen.getByTestId("epic-connection-pill").getAttribute("aria-label"),
+    ).toBe("All changes synced");
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx:77]: Covered by the same change in 57191ac8: the claim assertions go through `getByRole("button")` and its accessible name.

I kept the `data-testid` reads for the purely presentational assertions in that test (class names, inner markup, `data-status`) — those are asserting styling and the state attribute rather than the accessible surface, and a role query buys nothing there.

File: clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx [hunk 17]
```diff
@@ -269,23 +286,26 @@ describe("<EpicConnectionPill />", () => {
...
       rerender(pillTree());
 
+      // The LABEL still holds "Syncing…" (the anti-strobe trade), but the
+      // tooltip must not keep asserting a fact the verdict has already
+      // retracted. Every state now carries a tooltip, so the check is that the
+      // syncing copy is gone - not that the tooltip vanished.
       expect(screen.getByText("Syncing…")).not.toBeNull();
-      expect(pill().getAttribute("data-state")).toBeNull();
       fireEvent.focus(pill());
       expect(screen.queryByText(SYNCING_TOOLTIP)).toBeNull();
```
[the remark this answers is not in the document]
Comment (tanveergill) [reply clients/gui-app/src/components/epic-canvas/__tests__/epic-connection-pill.test.tsx]: Fixed in 57191ac8 — the settle-hold assertions now go through the `tooltip` role rather than `queryByText`.

One constraint worth recording: that test runs on fake timers, and `findByRole` polls on real timers, so it cannot settle there. It uses the synchronous `queryByRole("tooltip")` (Radix opens on focus in the same tick); the real-timer tests keep using the awaited `expectTooltip` helper.
